src/engine/Engine.h

```diff
@@ -367,8 +369,301 @@ class Engine {
+        if ((jcls_a & (1 << col)) == 0) {
+          res[col] = ID_NO_VALUE;
+        } else {
+          // if this is one of the join columns use the value in b
+          res[col] = (*b)[jclAToB[col]];
+        }
+        i++;
+      }
+      for (size_t col = 0; col < b->size(); col++) {
+        if ((jcls_b & (1 << col)) == 0) {
+          // only write the value if it is not one of the join columns in b
+          res[i] = (*b)[col];
+          i++;
+        }
+      }
+    } else if (bEmpty) {
+      // Fill the columns of b with ID_NO_VALUE and the rest with a
+      for (size_t col = 0; col < sizeA; col++) {
+        res[col] = (*a)[col];
+      }
+      for (size_t col = sizeA; col < resultSize; col++) {
+        res[col] = ID_NO_VALUE;
+      }
+    } else {
+      // Use the values from both a and b
+      unsigned int i = 0;
+      for (size_t col = 0; col < a->size(); col++) {
+        res[col] = (*a)[col];
+        i++;
+      }
+      for (size_t col = 0; col < b->size(); col++) {
+        if ((jcls_b & (1 << col)) == 0) {
+          res[i] = (*b)[col];
+          i++;
+        }
+      }
+    }
+  }
+
+  /**
+   * @brief Joins two result tables on any number of columns, inserting the
+   *        special value ID_NO_VALUE for any entries marked as optional.
+   * @param a
+   * @param b
+   * @param aOptional
+   * @param bOptional
+   * @param jcls
+   * @param result
+   */
+  template<typename A, typename B, typename R, int K>
+  static void optionalJoin(const A& a, const B& b,
+                           bool aOptional, bool bOptional,
+                           const vector<array<size_t, 2>>& jcls,
+                           vector<R> *result,
+                           unsigned int resultSize) {
+    // check for trivial cases
+    if ((a.size() == 0 && b.size() == 0)
+        || (a.size() == 0 && !aOptional)
+        || (b.size() == 0 && !bOptional)) {
+      return;
+    }
+
+    int jcls_a = 0;
+    int jcls_b = 0;
+    for (const array<size_t, 2>& jc : jcls) {
+      jcls_a |= (1 << jc[0]);
+      jcls_b |= (1 << jc[1]);
+    }
+
+    // When a is optional this is used to quickly determine
+    // in which column of b the value of a joined column can be found.
+    std::vector<size_t> jclAToB;
+    if (aOptional) {
+      uint32_t maxJoinColA = 0;
+      for (const array<size_t, 2>& jc : jcls) {
+        if (jc[0] > maxJoinColA) {
+          maxJoinColA = jc[0];
+        }
+      }
+      jclAToB.resize(maxJoinColA + 1);
+      for (const array<size_t, 2>& jc : jcls) {
+        jclAToB[jc[0]] = jc[1];
+      }
+    }
+
+    // Deal with one of the two tables beeing both empty and optional
+    if (a.size() == 0 && aOptional) {
+      size_t sizeA = resultSize - b[0].size() + jcls.size();
+      for (size_t ib = 0; ib < b.size(); ib++) {
+        R res = newOptionalResult<R, K>()(resultSize);
+        createOptionalResult<A, B, R, true, false>(
+              static_cast<typename A::value_type*>(0), &b[ib],
```

Comment:
`nullptr` should work without a cast, right?

Reply:
I originally relied on the compiler deducing the template arguments, thats why I had to cast the nullptrs to be pointers of the right type. Replaced them with actual nullptrs now.